lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,194 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+// Exceptions. / See https://llvm.org/LICENSE.txt for license information. /
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h.inc"
+#include "TPP/VNNIUtils.h"
+#include <iostream>
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimsForGemmLikeOp(RewriterBase &rewriter,
+                                    xsmm::BrgemmOp opTy) {
+
+  auto memrefC = opTy.getOperand(3).getType();
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+
+  int64_t m = memrefC.cast<ShapedType>().getShape()[0];
+  int64_t n = memrefC.cast<ShapedType>().getShape()[1];
+  int64_t k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb = (vnni::utils::isInVnniLayout(memrefB.cast<MemRefType>()))
+                    ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+                    : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+ArrayAttr getUnaryFlags(RewriterBase &rewriter) {
+  return rewriter.getArrayAttr(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+}
+
+ArrayAttr getBinaryFlags(RewriterBase &rewriter, xsmm::BinaryOp addOp) {
+  auto binaryInputType = addOp.getOperand(2).getType().cast<MemRefType>();
+  auto outputType = addOp.getOperand(3).getType().cast<MemRefType>();
+  return rewriter.getArrayAttr(xsmm::BinaryFlagsAttr::get(
+      rewriter.getContext(),
+      xsmm::utils::getBinaryBCast(binaryInputType, outputType,
+                                  /*operandNumber=*/0)));
+}
+
+xsmm::BinaryKindAttr getBinaryKind(RewriterBase &rewriter) {
```

Comment:
These helper functions return hardcoded attributes.
I'd give them more specific names to reflect that e.g., `getAddBinaryKind`.

Reply:
Honestly, this looks like a silly wrapper, totally unnecessary. Best to remove and use the call altogether.